Objects/regions.c

```diff
@@ -92,9 +116,324 @@ struct regionmetadata {
     regionmetadata* next;
 };
 
-static bool is_bridge_object(PyObject *op);
-static int regionmetadata_has_ancestor(regionmetadata* data, regionmetadata* other);
-static regionmetadata* PyRegion_get_metadata(PyRegionObject* obj);
+static Py_region_ptr_t regionmetadata_get_root(Py_region_ptr_t self)
+{
+    // Test for constants
+    if (IS_REGION_NULL(self) || IS_CONST_REGION(self)) {
+        return _Py_CAST(Py_region_ptr_t, self);
+    }
+
+    // Return self if it wasn't merged with another region
+    regionmetadata* self_data = _Py_CAST(regionmetadata*, self);
+    if (!REGION_PTR_GET_TAG(self_data->parent, Py_METADATA_MERGE_TAG)) {
+        return self;
+    }
+
+    // FIXME: It can happen that there are several layers in this union-find
+    // structure. It would be efficient to directly update the parent pointers
+    // for deeper nodes.
+    return regionmetadata_get_root(Py_region_ptr(self_data->parent));
+}
+#define regionmetadata_get_root(self) \
+    regionmetadata_get_root(_Py_CAST(Py_region_ptr_t, self))
+
+// This returns the represenative of this union-find structure or NULL
+// if the representative is a constant value.
+static regionmetadata* regionmetadata_get_root_or_null(Py_region_ptr_t self)
+{
+    Py_region_ptr_t rep = regionmetadata_get_root(self);
+    // Test for constants
+    if (IS_CONST_REGION(rep)) {
+        return NULL;
+    }
+
+    return _Py_CAST(regionmetadata*, rep);
+}
+#define regionmetadata_get_root_or_null(self) \
+    regionmetadata_get_root_or_null(_Py_CAST(Py_region_ptr_t, (self)))
+
+static void regionmetadata_open(regionmetadata* self) {
+    self = regionmetadata_get_root_or_null(self);
+    if (self) {
+        self->is_open = true;
+    }
+}
+
+static bool regionmetadata_is_open(regionmetadata* self) {
+    self = regionmetadata_get_root_or_null(self);
+    if (!self) {
+        // All constant regions are open
+        return true;
+    }
+
+    return self->is_open;
+}
+
+static void regionmetadata_inc_rc(Py_region_ptr_t self_raw)
+{
+    // Constant regions don't keep track of open subregions.
```

Comment:
I don't like the name "constant region" because this introduces a new concept which sound like it is on par with local region etc. but it is really just talking about how certain things are currently implemented. Would it be clearer to say e.g. "Regions without metadata objects"? (In multiple places!)

Reply:
(Phrase exists elsewhere but I only comment here.)